Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs

```diff
@@ -203,6 +333,324 @@ public static DualShockHIDOutputReport Create()
+                return false;
+        
+            var currentEvent = StateEvent.FromUnchecked(currentEventPtr);
+            var nextEvent = StateEvent.FromUnchecked(nextEventPtr);
+        
+            if (currentEvent->stateFormat != DualSenseHIDGenericInputReport.Format || nextEvent->stateFormat != DualSenseHIDGenericInputReport.Format)
+                return false;
+
+            if (currentEvent->stateSizeInBytes != nextEvent->stateSizeInBytes)
+                return false;
+
+            var currentGenericReport = (DualSenseHIDGenericInputReport*)currentEvent->state;
+            var nextGenericReport = (DualSenseHIDGenericInputReport*)nextEvent->state;
+
+            if (currentGenericReport->reportId != nextGenericReport->reportId)
+                return false;
+
+            if (currentGenericReport->reportId == 0x01)
+            {
+                if (currentEvent->stateSizeInBytes == 10 || currentEvent->stateSizeInBytes == 78)
+                {
+                    var currentState = (DualSenseHIDMinimalInputReport*)currentEvent->state;
+                    var nextState = (DualSenseHIDMinimalInputReport*)nextEvent->state;
+                    return MergeForward(currentState, nextState);
+                }
+                else
+                {
+                    var currentState = (DualSenseHIDUSBInputReport*)currentEvent->state;
+                    var nextState = (DualSenseHIDUSBInputReport*)nextEvent->state;
+                    return MergeForward(currentState, nextState);
+                }
+            }
+            else if (currentGenericReport->reportId == 0x31)
+            {
+                var currentState = (DualSenseHIDBluetoothInputReport*)currentEvent->state;
+                var nextState = (DualSenseHIDBluetoothInputReport*)nextEvent->state;
+                return MergeForward(currentState, nextState);
+            }
+            else
+                return false;
+        }
+        
+        unsafe bool IEventPreProcessor.PreProcessEvent(InputEventPtr eventPtr)
+        {
+            if (eventPtr.type != StateEvent.Type)
+                return eventPtr.type != DeltaStateEvent.Type; // only skip delta state events
+
+            var stateEvent = StateEvent.FromUnchecked(eventPtr);
+            var size = stateEvent->stateSizeInBytes;
+            if (stateEvent->stateFormat != DualSenseHIDGenericInputReport.Format || size < sizeof(DualSenseHIDInputReport))
+                return false; // skip unrecognized state events otherwise they will corrupt control states
+
+            var genericReport = (DualSenseHIDGenericInputReport*)stateEvent->state;
+            if (genericReport->reportId == 0x01)
+            {
+                if (stateEvent->stateSizeInBytes == 10 || stateEvent->stateSizeInBytes == 78)
+                {
+                    // minimal report
+                    var data = ((DualSenseHIDMinimalInputReport*)stateEvent->state)->ToHIDInputReport();
```

Comment:
Wondering if we need some validation here that what was returned from ToHIDInputReport isn't larger than the original size. Are there any guarantees that someone won't modify DualSenseHIDInputReport without realising this is a consequence?

Reply:
I've added a check and a test in IEventPreProcessor handler itself